packages/cli/src/controller/generate-controller.ts

```diff
@@ -0,0 +1,317 @@
+        function: fn.method,
+      },
+    };
+    formattedHandlers.push(handler);
+  });
+
+  userInput.events.map((event) => {
+    const handler: SubqlRuntimeHandler = {
+      handler: generateHandlerName(event.name, abiName, 'log'),
+      kind: EthereumHandlerKind.Event,
+      filter: {
+        topics: [event.method],
+      },
+    };
+    formattedHandlers.push(handler);
+  });
+
+  const assets = new Map<string, {file: string}>();
+  assets.set(abiName, {file: userInput.abiPath});
+
+  return {
+    kind: EthereumDatasourceKind.Runtime,
+    startBlock: userInput.startBlock,
+    options: {
+      abi: abiName,
+      address: userInput.address,
+    },
+    assets: assets,
+    mapping: {
+      file: DEFAULT_HANDLER_BUILD_PATH,
+      handlers: formattedHandlers,
+    },
+  };
+}
+
+// Selected fragments
+export async function prepareInputFragments<T extends ConstructorFragment | Fragment>(
+  type: 'event' | 'function',
+  input: string | undefined,
+  availableFragments: Record<string, T>,
+  abiName: string
+): Promise<Record<string, T>> {
+  // input = '' or undefined
+  if (input === undefined || input === '') {
+    return promptSelectables<T>(type, availableFragments);
+  }
+
+  // input = '*' (all Events/Functions)
+  if (input === '*') {
+    return availableFragments;
+  }
+
+  // input = 'transfer,approval'
+  const selectedFragments: Record<string, T> = {};
+  input.split(',').map((input) => {
+    const casedInput = input.trim().toLowerCase();
+    Object.keys(availableFragments).map((key) => {
+      if (casedInput === availableFragments[key].name.toLowerCase()) {
+        selectedFragments[key] = availableFragments[key];
+      }
+    });
+  });
+
+  // for (const key in availableFragments) {
+  //   if (inputs.includes(availableFragments[key].name.toLowerCase())) {
+  //     selectedFragments[key] = availableFragments[key];
+  //   }
+  //   //   console.log(chalk.red(`"${matchingInputs}" are duplicated on ABI: ${abiName}`));
+  // }
+  return selectedFragments;
+}
+
+export function filterExistingMethods(
+  eventFragments: Record<string, EventFragment>,
+  functionFragments: Record<string, FunctionFragment>,
+  dataSources: EthereumDs[]
+): [Record<string, EventFragment>, Record<string, FunctionFragment>] {
+  // find all existing functions and events
+  const existingEvents: string[] = [];
+  const existingFunctions: string[] = [];
+
+  const cleanEventsFragments: Record<string, EventFragment> = {};
+  const cleanFunctionsFragments: Record<string, FunctionFragment> = {};
+
+  dataSources.map((ds) => {
+    ds.mapping.handlers.map((handler) => {
+      if (Object.keys(handler.filter).includes('topics')) {
+        // topic[0] is the method
+        existingEvents.push((handler.filter as EthereumLogFilter).topics[0]);
+      }
+      if (Object.keys(handler.filter).includes('function')) {
+        existingFunctions.push((handler.filter as EthereumTransactionFilter).function);
+      }
+    });
+  });
```

Comment:
Does there need to be a check for a matching address? There could be an overlap with event/function name from a different contract.

It should be easy to write a test for this.